src/stroke.rs

```diff
@@ -0,0 +1,675 @@
+                    }
+                    Join::Miter => {
+                        if 2.0 * hypot < (hypot + dot) * style.miter_limit.powi(2) {
+                            // TODO: maybe better to store last_norm or derive from path?
+                            let last_scale = 0.5 * style.width / ab.hypot();
+                            let last_norm = last_scale * Vec2::new(-ab.y, ab.x);
+                            if cross > 0.0 {
+                                let fp_last = p0 - last_norm;
+                                let fp_this = p0 - norm;
+                                let h = ab.cross(fp_this - fp_last) / cross;
+                                let miter_pt = fp_this - cd * h;
+                                self.forward_path.line_to(miter_pt);
+                            } else if cross < 0.0 {
+                                let fp_last = p0 + last_norm;
+                                let fp_this = p0 + norm;
+                                let h = ab.cross(fp_this - fp_last) / cross;
+                                let miter_pt = fp_this - cd * h;
+                                self.backward_path.line_to(miter_pt);
+                            }
+                        }
+                        self.forward_path.line_to(p0 - norm);
+                        self.backward_path.line_to(p0 + norm);
+                    }
+                    Join::Round => {
+                        let angle = cross.atan2(dot);
+                        if cross > 0.0 {
+                            self.backward_path.line_to(p0 + norm);
+                            round_join(&mut self.forward_path, tolerance, p0, norm, angle);
+                        } else {
+                            self.forward_path.line_to(p0 - norm);
+                            round_join_rev(&mut self.backward_path, tolerance, p0, -norm, -angle);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn do_line(&mut self, style: &Stroke, tangent: Vec2, p1: Point) {
+        let scale = 0.5 * style.width / tangent.hypot();
+        let norm = scale * Vec2::new(-tangent.y, tangent.x);
+        self.forward_path.line_to(p1 - norm);
+        self.backward_path.line_to(p1 + norm);
+        self.last_pt = p1;
+    }
+
+    fn do_cubic(&mut self, style: &Stroke, c: CubicBez, tolerance: f64, opts: &StrokeOpts) {
+        let co = CubicOffset::new(c, -0.5 * style.width);
+        let forward = fit_with_opts(&co, tolerance, opts);
+        self.forward_path.extend(forward.into_iter().skip(1));
+        // A tuning parameter for regularization. A value too large may distort the curve,
+        // while a value too small may fail to generate smooth curves. This is a somewhat
+        // arbitrary value, and should be revisited.
+        const DIM_TUNE: f64 = 0.25;
+        let co = CubicOffset::new_regularized(c, 0.5 * style.width, tolerance * DIM_TUNE);
```

Comment:
I’m confused as to why is this done here and only for the backward path? In my testing version, I called regularize on the initial cubic itself in stroke_undashed before computing the tangents for joins. My assumption was that regularize fixes zero length tangent vectors and we would want consistency there.

Reply:
Simple brain fart error. I should have validated this more carefully before pushing. Thanks for the catch!